File: lib/src/web/readers/messages/event_message_reader.dart
// Copyright (c) 2022-present Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

import 'package:flutter/foundation.dart';
import '../entities/media_ad_break_entity_reader.dart';
import '../entities/media_ad_entity_reader.dart';
import '../entities/media_player_entity_reader.dart';
import '../events/media_event_reader.dart';
import '../events/consent_granted_reader.dart';
import '../events/consent_withdrawn_reader.dart';
import '../events/contexts_reader.dart';
import '../events/event_reader.dart';
import '../events/screen_view_reader.dart';
import '../events/self_describing_reader.dart';
import '../events/structured_reader.dart';
import '../events/timing_reader.dart';
import '../events/page_view_event_reader.dart';
import '../events/scroll_changed_reader.dart';
import '../events/list_item_view_reader.dart';

@immutable
class EventMessageReader {
  final String tracker;
  final EventReader event;
  final ContextsReader? contexts;
  final String? mediaTrackingId;
  final MediaPlayerEntityReader? player;
  final MediaAdEntityReader? ad;
  final MediaAdBreakEntityReader? adBreak;

  const EventMessageReader(
      {required this.tracker,
      required this.event,
      this.contexts,
      this.mediaTrackingId,
      this.player,
      this.ad,
      this.adBreak});

  EventMessageReader.withStructured(dynamic map)
      : tracker = map['tracker'],
        event = StructuredReader(map['eventData']),
        contexts =
            map['contexts'] != null ? ContextsReader(map['contexts']) : null,
        mediaTrackingId = null,
        player = null,
        ad = null,
        adBreak = null;

  EventMessageReader.withSelfDescribing(dynamic map)
      : tracker = map['tracker'],
        event = SelfDescribingReader(map['eventData'],
            isMedia: map['mediaTrackingId'] != null),
        contexts =
            map['contexts'] != null ? ContextsReader(map['contexts']) : null,
        mediaTrackingId = map['mediaTrackingId'],
        player = map['player'] != null
            ? MediaPlayerEntityReader(map['player'])
            : null,
        ad = map['ad'] != null ? MediaAdEntityReader(map['ad']) : null,
        adBreak = map['adBreak'] != null
            ? MediaAdBreakEntityReader(map['adBreak'])
            : null;

  EventMessageReader.withScreenView(dynamic map)
      : tracker = map['tracker'],
        event = ScreenViewReader(map['eventData']),
        contexts =
            map['contexts'] != null ? ContextsReader(map['contexts']) : null,
        mediaTrackingId = null,
        player = null,
        ad = null,
        adBreak = null;

  EventMessageReader.withScrollChanged(dynamic map)
      : tracker = map['tracker'],
        event = ScrollChangedReader(map['eventData']),
        contexts =
            map['contexts'] != null ? ContextsReader(map['contexts']) : null,
        mediaTrackingId = null,
        player = null,
        ad = null,
        adBreak = null;

  EventMessageReader.withListItemView(dynamic map)
      : tracker = map['tracker'],
        event = ListItemViewReader(map['eventData']),
        contexts =
            map['contexts'] != null ? ContextsReader(map['contexts']) : null,
        mediaTrackingId = null,
        player = null,
        ad = null,
        adBreak = null;

  EventMessageReader.withTiming(dynamic map)
      : tracker = map['tracker'],
        event = TimingReader(map['eventData']),
        contexts =
            map['contexts'] != null ? ContextsReader(map['contexts']) : null,
        mediaTrackingId = null,
        player = null,
        ad = null,
        adBreak = null;

  EventMessageReader.withConsentGranted(dynamic map)
      : tracker = map['tracker'],
        event = ConsentGrantedReader(map['eventData']),
        contexts =
            map['contexts'] != null ? ContextsReader(map['contexts']) : null,
        mediaTrackingId = null,
        player = null,
        ad = null,
        adBreak = null;

  EventMessageReader.withConsentWithdrawn(dynamic map)
      : tracker = map['tracker'],
        event = ConsentWithdrawnReader(map['eventData']),
        contexts =
            map['contexts'] != null ? ContextsReader(map['contexts']) : null,
        mediaTrackingId = null,
        player = null,
        ad = null,
        adBreak = null;

  EventMessageReader.withPageView(dynamic map)
      : tracker = map['tracker'],
        event = PageViewEventReader(map['eventData']),
        contexts =
            map['contexts'] != null ? ContextsReader(map['contexts']) : null,
        mediaTrackingId = null,
        player = null,
        ad = null,
        adBreak = null;

  EventMessageReader.withMediaEvent(MediaEndpoint mediaEndpoint, dynamic map)
      : tracker = map['tracker'],
        event = MediaEventReader(
            mediaEndpoint: mediaEndpoint, data: map['eventData']),
        contexts =
            map['contexts'] != null ? ContextsReader(map['contexts']) : null,
        mediaTrackingId = map['mediaTrackingId'],
        player = map['player'] != null
            ? MediaPlayerEntityReader(map['player'])
            : null,
        ad = map['ad'] != null ? MediaAdEntityReader(map['ad']) : null,
        adBreak = map['adBreak'] != null
            ? MediaAdBreakEntityReader(map['adBreak'])
            : null;

  dynamic eventData() {
    dynamic data = event.eventData();
    if (contexts != null) {
      data['context'] =
          contexts?.selfDescribingJsons.map((e) => e.json()).toList();
    }
    if (mediaTrackingId != null) {
      data['id'] = mediaTrackingId;
    }
    if (player != null) {
      data['player'] = player?.toMap();
    }
    if (ad != null) {
      data['ad'] = ad?.toMap();
    }
    if (adBreak != null) {
      data['adBreak'] = adBreak?.toMap();
    }
    return data;
  }
}

File: android/src/main/kotlin/com/snowplowanalytics/snowplow_tracker/readers/messages/EventMessageReader.kt
// Copyright (c) 2022-present Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

package com.snowplowanalytics.snowplow_tracker.readers.messages

import com.snowplowanalytics.snowplow.event.*
import com.snowplowanalytics.snowplow.media.event.*
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson
import com.snowplowanalytics.snowplow_tracker.readers.entities.MediaAdBreakEntityReader
import com.snowplowanalytics.snowplow_tracker.readers.entities.MediaAdEntityReader
import com.snowplowanalytics.snowplow_tracker.readers.entities.MediaPlayerEntityReader
import com.snowplowanalytics.snowplow_tracker.readers.events.*

class EventMessageReader(val values: Map<String, Any>) {
    private val valuesDefault = values.withDefault { null }

    val tracker: String by values
    private val eventData: Map<String, Any> by values
    private val contexts: List<Map<String, Any>>? by valuesDefault
    private val contextsJsons: List<SelfDescribingJson>? by lazy {
        contexts?.let { it.map { item -> SelfDescribingJsonReader(item).toSelfDescribingJson() } }
    }
    val mediaTrackingId: String? by valuesDefault
    val player: MediaPlayerEntityReader? by lazy {
        values.get("player")?.let {
            MediaPlayerEntityReader(it as Map<String, Any>)
        }
    }
    val ad: MediaAdEntityReader? by lazy {
        values.get("ad")?.let {
            MediaAdEntityReader(it as Map<String, Any>)
        }
    }
    val adBreak: MediaAdBreakEntityReader? by lazy {
        values.get("adBreak")?.let {
            MediaAdBreakEntityReader(it as Map<String, Any>)
        }
    }

    fun toStructuredWithContexts(): Structured {
        val structured = StructuredReader(eventData).toStructured()
        addContext(structured)
        return structured
    }

    fun toSelfDescribingWithContexts(): SelfDescribing {
        val selfDescribing = SelfDescribingJsonReader(eventData).toSelfDescribing()
        addContext(selfDescribing)
        return selfDescribing
    }

    fun toScreenViewWithContexts(): ScreenView {
        val screenView = ScreenViewReader(eventData).toScreenView()
        addContext(screenView)
        return screenView
    }

    fun toScrollChangedWithContexts(): ScrollChanged {
        val scrollChanged = ScrollChangedReader(eventData).toScrollChanged()
        addContext(scrollChanged)
        return scrollChanged
    }

    fun toListItemViewWithContexts(): ListItemView {
        val listItemView = ListItemViewReader(eventData).toListItemView()
        addContext(listItemView)
        return listItemView
    }

    fun toTimingWithContexts(): Timing {
        val timing = TimingReader(eventData).toTiming()
        addContext(timing)
        return timing
    }

    fun toConsentGrantedWithContexts(): ConsentGranted {
        val consentGranted = ConsentGrantedReader(eventData).toConsentGranted()
        addContext(consentGranted)
        return consentGranted
    }

    fun toConsentWithdrawnWithContexts(): ConsentWithdrawn {
        val consentWithdrawn = ConsentWithdrawnReader(eventData).toConsentWithdrawn()
        addContext(consentWithdrawn)
        return consentWithdrawn
    }

    fun toMediaAdBreakEndEvent(): MediaAdBreakEndEvent {
        val event = MediaAdBreakEndEvent()
        addContext(event)
        return event
    }

    fun toMediaAdBreakStartEvent(): MediaAdBreakStartEvent {
        val event = MediaAdBreakStartEvent()
        addContext(event)
        return event
    }

    fun toMediaAdClickEvent(): MediaAdClickEvent {
        val event = MediaAdEventReader(eventData).toMediaAdClickEvent()
        addContext(event)
        return event
    }

    fun toMediaAdCompleteEvent(): MediaAdCompleteEvent {
        val event = MediaAdCompleteEvent()
        addContext(event)
        return event
    }

    fun toMediaAdFirstQuartileEvent(): MediaAdFirstQuartileEvent {
        val event = MediaAdFirstQuartileEvent()
        addContext(event)
        return event
    }

    fun toMediaAdMidpointEvent(): MediaAdMidpointEvent {
        val event = MediaAdMidpointEvent()
        addContext(event)
        return event
    }

    fun toMediaAdPauseEvent(): MediaAdPauseEvent {
        val event = MediaAdEventReader(eventData).toMediaAdPauseEvent()
        addContext(event)
        return event
    }

    fun toMediaAdResumeEvent(): MediaAdResumeEvent {
        val event = MediaAdEventReader(eventData).toMediaAdResumeEvent()
        addContext(event)
        return event
    }

    fun toMediaAdSkipEvent(): MediaAdSkipEvent {
        val event = MediaAdEventReader(eventData).toMediaAdSkipEvent()
        addContext(event)
        return event
    }

    fun toMediaAdStartEvent(): MediaAdStartEvent {
        val event = MediaAdStartEvent()
        addContext(event)
        return event
    }

    fun toMediaAdThirdQuartileEvent(): MediaAdThirdQuartileEvent {
        val event = MediaAdThirdQuartileEvent()
        addContext(event)
        return event
    }

    fun toMediaBufferEndEvent(): MediaBufferEndEvent {
        val event = MediaBufferEndEvent()
        addContext(event)
        return event
    }

    fun toMediaBufferStartEvent(): MediaBufferStartEvent {
        val event = MediaBufferStartEvent()
        addContext(event)
        return event
    }

    fun toMediaEndEvent(): MediaEndEvent {
        val event = MediaEndEvent()
        addContext(event)
        return event
    }

    fun toMediaErrorEvent(): MediaErrorEvent {
        val event = MediaErrorEventReader(eventData).toMediaErrorEvent()
        addContext(event)
        return event
    }

    fun toMediaFullscreenChangeEvent(): MediaFullscreenChangeEvent {
        val event = MediaFullscreenChangeEventReader(eventData).toMediaFullscreenChangeEvent()
        addContext(event)
        return event
    }

    fun toMediaPauseEvent(): MediaPauseEvent {
        val event = MediaPauseEvent()
        addContext(event)
        return event
    }

    fun toMediaPictureInPictureChangeEvent(): MediaPictureInPictureChangeEvent {
        val event = MediaPictureInPictureChangeEventReader(eventData).toMediaPictureInPictureChangeEvent()
        addContext(event)
        return event
    }

    fun toMediaPlayEvent(): MediaPlayEvent {
        val event = MediaPlayEvent()
        addContext(event)
        return event
    }

    fun toMediaPlaybackRateChangeEvent(): MediaPlaybackRateChangeEvent {
        val event = MediaPlaybackRateChangeEventReader(eventData).toMediaPlaybackRateChangeEvent()
        addContext(event)
        return event
    }

    fun toMediaQualityChangeEvent(): MediaQualityChangeEvent {
        val event = MediaQualityChangeEventReader(eventData).toMediaQualityChangeEvent()
        addContext(event)
        return event
    }

    fun toMediaReadyEvent(): MediaReadyEvent {
        val event = MediaReadyEvent()
        addContext(event)
        return event
    }

    fun toMediaSeekEndEvent(): MediaSeekEndEvent {
        val event = MediaSeekEndEvent()
        addContext(event)
        return event
    }

    fun toMediaSeekStartEvent(): MediaSeekStartEvent {
        val event = MediaSeekStartEvent()
        addContext(event)
        return event
    }

    fun toMediaVolumeChangeEvent(): MediaVolumeChangeEvent {
        val event = MediaVolumeChangeEventReader(eventData).toMediaVolumeChangeEvent()
        addContext(event)
        return event
    }

    private fun addContext(event: Event) {
        contextsJsons?.let { event.entities.addAll(it) }
    }
}
